source/common/common/inline_map_registry.h

```diff
@@ -0,0 +1,389 @@
+ *   // than all inline maps created by the registry.
+ *   InlineMapRegistry registry;
+ *
+ *   // Register the inline key. We should never do this after bootstrapping. Typically, we can
+ *   // do this by define a global variable.
+ *   auto inline_handle = registry.registerInlineKey("inline_key");
+ *
+ *   // Finalize the registry. No further changes are allowed to the registry after this point.
+ *   registry.finalize();
+ *
+ *   // Create the inline map.
+ *   auto inline_map = registry.createInlineMap<std::string>();
+ *
+ *   // Get value by inline handle.
+ *   inline_map->insert(inline_handle, std::make_unique<std::string>("value"));
+ *   EXPECT_EQ(*inline_map->lookup(inline_handle), "value");
+ */
+class InlineMapRegistry {
+public:
+  /**
+   * This is the handle used to access the key/value pair in the inline map. The handle is
+   * lightweight and could be copied and passed by value.
+   */
+  class InlineKey {
+  public:
+    /**
+     * Get the id of the inline map registry that the handle created by. This will be used
+     * to determine if the handle could be accepted by a inline map.
+     */
+    uint64_t registryId() const { return registry_id_; }
+
+    /**
+     * Get the id of the inline entry in the inline map. This could be used to access the
+     * key/value pair in the inline map.
+     */
+    uint64_t inlineId() const { return inline_id_; }
+
+    /**
+     * Get the key view of the inline key.
+     */
+    absl::string_view inlineKey() const { return inline_key_; }
+
+  private:
+    friend class InlineMapRegistry;
+
+    // This constructor should only be called by InlineMapRegistry.
+    InlineKey(uint64_t registry_id, uint64_t inline_id, absl::string_view inline_key)
+        : registry_id_(registry_id), inline_id_(inline_id), inline_key_(inline_key) {}
+
+    uint64_t registry_id_{};
+    uint64_t inline_id_{};
+    absl::string_view inline_key_;
+  };
+
+  // This is the Hash used for registration map/set and underlay hash map.
+  using Hash = absl::container_internal::hash_default_hash<std::string>;
+
+  using RegistrationMap = absl::flat_hash_map<absl::string_view, uint64_t, Hash>;
+
+  // Set to store the inline keys. Because the node of the set have stable address, we can reference
+  // the key in the handle or registration map safely.
+  using RegistrationSet = absl::node_hash_set<std::string, Hash>;
+
+  template <class T> class InlineMap : public InlineStorage {
+  public:
+    using TPtr = std::unique_ptr<T>;
+    using UnderlayHashMap = absl::flat_hash_map<std::string, TPtr, Hash>;
+
+    // Get the entry for the given key. If the key is not found, return nullptr.
+    const T* lookup(absl::string_view key) const { return lookupImpl(key); }
+    T* lookup(absl::string_view key) { return lookupImpl(key); }
+
+    // Get the entry for the given handle. If the handle is not valid, return nullptr.
+    const T* lookup(InlineKey handle) const { return lookupImpl(handle); }
```

Comment:
might as well use a different names for LookupImp rather than overloading at that layer, e.g. lookupInline vs lookupDynamic.

Reply:
I think this overloading bring lots convenience. orz Esp when we implement some wrapper class. You can check the filter state impl to see this overloading could reduce lots of unnecessary code.